### ISSUE TITLE
Jenkins-SFDX retrieve failing using freestyle projects. No issue with pipeline projects.

### ISSUE THREAD

**Issue by cumulus-robert** [forcedotcom/cli#433]
### Summary

Unable to use Freeform Projects in Jenkins, only groovy scripts.

### Steps To Reproduce:

* Create a freestyle Project. Add shell script to retrieve. Review Output.

### Expected result

* It should retrieve per the script

### Actual result

* It fails with this reason:

ERROR running force:auth:jwt:grant:  We encountered a JSON web token error, which is likely not an issue with Salesforce CLI. Here’s the error: EISDIR: illegal operation on a directory, read
Build step 'Execute shell' marked build as failure
Build did not succeed and the project is configured to only push after a successful build, so no pushing will occur.
Finished: FAILURE

### Additional information

I have tried what I know to try. I've:
* reinstalled SFDX
* perform chown on sfdx related folders to all users I can think of impacted: root, jenkins, my local user. All return same result.

Specifically this is the command I was using:

chown -R username:username /home/%user of install or sudo update%/.local/share/sfdx/
OR
chown -R username /home/%user of install or sudo update%/.local/share/sfdx/

Where username was replaced with each of the above. 

**SFDX CLI Version**(to find the version of the CLI engine run sfdx --version):

sfdx-cli/7.61.0-af69066d0d linux-x64 node-v10.20.1

**SFDX plugin Version**(to find the version of the CLI plugin run sfdx plugins --core)
@oclif/plugin-autocomplete 0.1.5 (core)
@oclif/plugin-commands 1.2.3 (core)
@oclif/plugin-help 2.2.3 (core)
@oclif/plugin-not-found 1.2.3 (core)
@oclif/plugin-plugins 1.7.9 (core)
@oclif/plugin-update 1.3.9 (core)
@oclif/plugin-warn-if-update-available 1.7.0 (core)
@oclif/plugin-which 1.0.3 (core)
@salesforce/sfdx-trust 3.0.7 (core)
analytics 1.7.1 (core)
generator 1.1.2 (core)
salesforcedx 48.17.0 (core)
├─ @salesforce/sfdx-plugin-lwc-test 0.1.5 (core)
├─ salesforcedx-templates 48.18.0 (core)
└─ salesforce-alm 48.18.0 (core)
sfdx-cli 7.61.0 (core)

I should note this is an out of box deployment/install of Jenkins. Freestyle projects have never worked since deployment. Only groovy scripts.

**OS and version**:

Debian GNU/Linux 9 (stretch) on Google Console Bitnami VM

**Comment by clairebianchi** [comment]
@cumulus-robert can you please provide a repo where this is happening? We do not use freestyle projects but do use Jenkins for our builds and are able to auth via JWT without a problem.

**Comment by cumulus-robert** [comment]
@Clairebianchi –

 

I actually just got this working this Tuesday. I unfortunately couldn’t spend time troubleshooting this when this started in March this year. 

 

Our issues still came down to SFDX authentication procedures changing in Q4 last year unfortunately and the adverse impact that change had to our Jenkins instance.

 

I think it would be helpful to include information in your Jenkins walkthrough around this specifically. I re-read the documentation about using NPM for CI installations, but that wasn’t sufficient. I followed the documentation every word and still experienced permissions issues. Since Jenkins runs as a service, SFDX was still running as sudo root (even though script wasn’t a sudo command) and accessing root folders that the Jenkins user doesn’t have access to. The moment any SFDX command was attempted, a cache folder was attempting to be added to a folder that Jenkins user doesn’t have access to. The feedback in the documentation suggesting to overcome this, doesn’t work (installing SFDX using NPM and NodeJS). Additional steps are needed in order for freestyle projects to work.

 

Don’t ask me why our pipelines worked just fine, that was my only saving grace since March. I could still run deployments using pipeline groovy scrips, I just couldn’t do automatic merges, automatic conflict checks ahead of any CI activity. I do these activities in Jenkins using Jenkins free-style projects.

 

I am running Jenkins on ubuntu and this is what I did to solve the permissions issue:

 

Install ACL (if not already installed)

 

sudo apt-get install acl

 

Then update folder permissions (this folder is prompted in SFDX error)

 

sudo su -
cd /
sudo chown -R jenkins root
sudo setfacl -R -m user:jenkins:rwx root
 

This gives Jenkins user folder ownership and RW access to the folder that SFDX is attempting to add cache folder to.

 

Robert
